src/chain/chain-supervision.cc

```diff
@@ -603,7 +683,8 @@ int32 ComputeFstStateTimes(const fst::StdVectorFst &fst,
 Supervision::Supervision(const Supervision &other):
     weight(other.weight), num_sequences(other.num_sequences),
     frames_per_sequence(other.frames_per_sequence),
-    label_dim(other.label_dim), fst(other.fst) { }
+    label_dim(other.label_dim), fst(other.fst),
+    e2e(other.e2e), e2e_fsts(other.e2e_fsts) { }
 
 void AppendSupervision(const std::vector<const Supervision*> &input,
                        bool compactify,
```

Comment:
These functions AppendSupervision() and AddWeightToSupervisionFst() have become very long due to this e2e stuff.  We try to avoid having functions longer than about 40 lines.  I think the best way to keep the two parts of the code separate would be to have a statement near the beginning of those functions that checks if the 'e2e' flag is set, and if it is, recurses to a static function with a name like `AppendSupervisionE2e()`.